File: gtracr/lib/src/TrajectoryPoint.cc
/*
    Class that records a single point in the particle trajectory
    Members:
    - latitude: the geographic latitude, with 0 at the equator (+90 at the North Pole, -90 at South pole) in degrees
    - longitude: the geographic longitude, with 0 at the Prime Meridian (negative degrees towards Western hemisphere) in degrees
    - altitude: the altitude above sea level
    - v_r: the radial velocity
    - v_theta: the velocity in the polar direction
    - v_phi: the velocity in the azimuthal direction
*/

// #include <array>
#include <vector>
#include <iostream>
#include <math.h>
#include "constants.h"
#include "Matrix.h"
#include "Location.h"
#include "TrajectoryPoint.h"

// Constructors
// default constructor
TrajectoryPoint::TrajectoryPoint()
    :lat{0.}, lng{0.}, alt{0.}, v_r{0.}, v_theta{0.}, v_phi{0.}
{

}
// variable location and velocity
TrajectoryPoint::TrajectoryPoint(const double &latitude, const double &longitude, const double &altitude, const double &rvel, const double &thvel, const double &phvel)
    : lat{latitude}, lng{longitude}, alt{altitude}, v_r{rvel}, v_theta{thvel}, v_phi{phvel}
{
}

// initialize with Location object
TrajectoryPoint::TrajectoryPoint(const Location &loc, const double &rvel, const double &thvel, const double &phvel)
    : v_r{rvel}, v_theta{thvel}, v_phi{phvel}
{
    lat = loc.latitude();
    lng = loc.longitude();
    alt = loc.altitude();
}

// copy constructor
TrajectoryPoint::TrajectoryPoint(const TrajectoryPoint &TJP)
    : lat{TJP.lat}, lng{TJP.lng}, alt{TJP.alt}, v_r{TJP.v_r}, v_theta{TJP.v_theta}, v_phi{TJP.v_phi}
{
}

// copy assignment operator
TrajectoryPoint &TrajectoryPoint::operator=(const TrajectoryPoint &TJP)
{
    lat = TJP.lat;
    lng = TJP.lng;
    alt = TJP.alt;
    v_r = TJP.v_r;
    v_theta = TJP.v_theta;
    v_phi = TJP.v_phi;
    return *this;
}

// get the magnitude of the coordinate vector in geocentric coordinates
double TrajectoryPoint::magnitude()
{
    const double &mag = alt - constants::RE;
    return mag;
}

// get the magnitude of the velocity in geocentric coordinates
double TrajectoryPoint::velocity()
{
    std::vector<double> sphvec = getSphericalCoord();
    double r = sphvec[0];
    double theta = sphvec[1];
    double phi = sphvec[2];
    // sphvec[0] = r, sphvec[1] = theta, sphvec[2] = phi
    double vel = sqrt(v_r * v_r + r * r * v_theta * v_theta + r * r * sin(phi) * sin(phi) * v_phi * v_phi);
    return vel;
}

// convert geodesic coordinates to spherical coordinates
std::vector<double> TrajectoryPoint::getSphericalCoord()
{
    double r = constants::RE + alt;
    double theta = (90. - lat) * constants::DEG_TO_RAD;
    double phi = lng * constants::DEG_TO_RAD;
    std::vector<double> sphcoordarr = {r, theta, phi};
    // std::vector<double> sphcoordvec = std::vector<double>(0, 3, sphcoordarr);
    return sphcoordarr;
}

// convert spherical coordinates to geodesic coordinates
void TrajectoryPoint::setSphericalCoord(const double &r, const double &theta, const double &phi)
{
    lat = 90. - (theta * constants::RAD_TO_DEG);
    lng = phi * constants::RAD_TO_DEG;
    alt = r - constants::RE;
}

// convert geodesic to cartesian
std::vector<double> TrajectoryPoint::getCartesianCoord()
{
    double lmbda = lat * constants::DEG_TO_RAD;
    double eta = lng * constants::DEG_TO_RAD;
    double x = (constants::RE + alt) * cos(lmbda) * cos(eta);
    double y = (constants::RE + alt) * cos(lmbda) * sin(eta);
    double z = (constants::RE + alt) * sin(lmbda);
    std::vector<double> carcoordarr = {x, y, z};
    // std::cout << carcoordarr[0] << carcoordarr[1] << std::endl;
    // std::vector<double> carcoordvec = std::vector<double>(0, 3, carcoordarr);
    // carcoordvec.print();
    return carcoordarr;
}

// convert cartesian to geodesic
// void TrajectoryPoint::setCartesianCoord(const double &x, const double &y, const double &z)
// {

// }

// convert spherical velocity to cartesian velocity
std::vector<double> TrajectoryPoint::getCartesianVelocity()
{
    std::vector<double> sphvec = getSphericalCoord();
    double r = sphvec[0];
    double theta = sphvec[1];
    double phi = sphvec[2];
    // sphvec[0] = r, sphvec[1] = theta, sphvec[2] = phi
    double vx = v_r * sin(theta) * cos(phi) + r * v_theta * cos(theta) * cos(phi) - r * v_phi * sin(theta) * sin(phi);
    double vy = v_r * sin(theta) * sin(phi) + r * v_theta * cos(theta) * sin(phi) + r * v_phi * sin(theta) * cos(phi);
    double vz = v_r * cos(theta) - r * v_theta * sin(theta);
    std::vector<double> carvelarr = {vx, vy, vz};
    // std::vector<double> carvelvec = std::vector<double>(0, 3, carvelarr);
    return carvelarr;
}

// convert cartesian velocity to spherical velocity
void TrajectoryPoint::setCartesianVelocity(const double &vx, const double &vy, const double &vz)
{
    std::vector<double> sphvec = getSphericalCoord();
    double r = sphvec[0];
    double theta = sphvec[1];
    double phi = sphvec[2];
    // sphvec[0] = r, sphvec[1] = theta, sphvec[2] = phi
    v_r = vx * sin(theta) * cos(phi) + vy * sin(theta) * sin(phi) + vz * cos(theta);
    v_theta = (vx * cos(theta) * cos(phi) + vy * cos(theta) * sin(phi) - vz * sin(theta)) / r;
    v_phi = (-vx * sin(phi) + vy * cos(phi)) / (r * sin(theta));
}

// get the spherical components (coordinate and velocity)
std::vector<double> TrajectoryPoint::spherical()
{
    std::vector<double> sphcoordvec = getSphericalCoord();
    // sphvec[0] = r, sphvec[1] = theta, sphvec[2] = phi
    std::vector<double> sphvecarr = {sphcoordvec[0], sphcoordvec[1], sphcoordvec[2], v_r, v_theta, v_phi};
    // std::vector<double> sphvec = std::vector<double>(0, 6, sphvecarr);
    return sphvecarr;
}

// get the cartesian components (coordinate and velocity)
std::vector<double> TrajectoryPoint::cartesian()
{
    std::vector<double> carcoordvec = getCartesianCoord();
    std::vector<double> carvelvec = getCartesianVelocity();
    std::vector<double> carvecarr = {carcoordvec[0], carcoordvec[1], carcoordvec[2], carvelvec[0], carvelvec[1], carvelvec[2]};
    // std::vector<double> carvec = std::vector<double>(0, 6, carvecarr);
    return carvecarr;
}

// equivalent to __str__ in Python
// print the class object
void TrajectoryPoint::print()
{
    std::cout << "Latitude: " << lat << ", "
              << "Longitude: " << lng << ", "
              << "Altitude: " << alt << ", "
              << "Velocity (vr, vtheta, vphi): "
              << "( " << v_r << ", "
              << v_theta << ", "
              << v_phi << " )"
              << std::endl;
}

File: gtracr/trajectorypoint.py
import os, sys
import numpy as np

# sys.path.append(os.getcwd())
# sys.path.append(os.path.join(os.getcwd(), "gtracr"))

from gtracr.constants import EARTH_RADIUS, DEG_TO_RAD, RAD_TO_DEG

# KEY_LIST = ["t", "r", "theta", "phi", "pr", "ptheta", "pphi"]


class TrajectoryPoint:
    '''
    Class that records a single point in the particle trajectory
    Members:
    - r: the radial component [m]
    - theta: the polar component, with 0 defined at the North Pole
    - phi: the azimuthal component, with 0 defined at the Prime Meridian (domain: [-pi, pi])
    - pr: the radial momentum
    - ptheta: the momentum in the polar direction
    - pphi: the momentum in the azimuthal direction
    '''
    def __init__(self,
                 r=EARTH_RADIUS,
                 theta=0.,
                 phi=0.,
                 pr=0.,
                 ptheta=0.,
                 pphi=0.):
        self.r = r
        self.theta = theta
        self.phi = phi
        self.pr = pr
        self.ptheta = ptheta
        self.pphi = pphi

    # get geodesic coordinate equivalents of spherical ones
    def geodesic_coordinate(self):
        latitude = 90. - (self.theta * RAD_TO_DEG)
        # longitude = (phi * RAD_TO_DEG) - 180.
        longitude = self.phi * RAD_TO_DEG
        altitude = self.r - EARTH_RADIUS
        return np.array([latitude, longitude, altitude])

    # set latitude, longitude, altitude from geodesic coordiantes
    def set_geodesic_coord(self, latitude, longitude, altitude):
        self.r = EARTH_RADIUS + altitude
        self.theta = (
            90. - latitude
        ) * DEG_TO_RAD  # theta defined in [0, pi], theta = 90 at equator
        self.phi = longitude * DEG_TO_RAD  # phi defined in [-pi, pi], phi = 0 at prime meridian

    # get cartesian coordinate equivalents of spherical ones
    def cartesian_coord(self):
        x = self.r * np.sin(self.theta) * np.cos(self.phi)
        y = self.r * np.sin(self.theta) * np.sin(self.phi)
        z = self.r * np.cos(self.theta)

        return np.array([x, y, z])

    # set latitude, longitude, altitude from Cartesian coordiantes
    def set_cartesian_coord(self, x, y, z):
        self.r = np.sqrt(x**2. + y**2. + z**2.)
        self.theta = np.arccos(z / self.r)
        self.phi = np.arctan2(y, x)

    # set pr, ptheta, pphi from cartesian momentum
    def set_cartesian_momentum(self, px, py, pz):
        self.pr = px * np.sin(self.theta) * np.cos(self.phi) + py * np.sin(
            self.theta) * np.sin(self.phi) + pz * np.cos(self.theta)
        self.ptheta = (px * np.cos(self.theta) * np.cos(self.phi) +
                       py * np.cos(self.theta) * np.sin(self.phi) -
                       pz * np.sin(self.theta))
        self.pphi = (-px * np.sin(self.phi) + py * np.cos(self.phi))

        # return np.array([vr, vtheta, vphi])

    # return cartesian momenta from spherical ones
    def cartesian_momentum(self):
        # px = self.pr * np.sin(self.theta) * np.cos(
        #     self.phi) + self.r * self.ptheta * np.cos(self.theta) * np.cos(
        #         self.phi) - self.r * self.pphi * np.sin(self.theta) * np.sin(
        #             self.phi)
        # py = self.pr * np.sin(self.theta) * np.sin(
        #     self.phi) + self.r * self.ptheta * np.cos(self.theta) * np.sin(
        #         self.phi) + self.r * self.pphi * np.sin(self.theta) * np.cos(
        #             self.phi)
        # pz = self.pr * np.cos(self.theta) - self.r * self.ptheta * np.sin(
        #     self.theta)

        tfmat_sph2car = np.array([[
            [
                np.sin(self.theta) * np.cos(self.phi),
                self.r * np.cos(self.theta) * np.cos(self.phi),
                -self.r * np.sin(self.theta) * np.sin(self.phi)
            ],
            [
                np.sin(self.theta) * np.sin(self.phi),
                self.r * np.cos(self.theta) * np.sin(self.phi),
                self.r * np.sin(self.theta) * np.cos(self.phi),
            ],
            [np.cos(self.theta), -self.r * np.sin(self.theta), 0.],
        ]])

        cart_momentum_arr = np.dot(tfmat_sph2car,
                                   np.array([self.pr, self.ptheta, self.pphi]))

        return cart_momentum_arr

    def __str__(self):
        return "Coordinate (r, theta, phi): ({:.5e}, {:.5e}, {:.5e}), \
            \nMomentum (pr, ptheta, pphi): ({:.5e}, {:.5e}, {:.5e})\n".format(
            *tuple(vars(self).values()))

    # both position and momentum in Cartesian coordinates
    def cartesian(self):
        return np.concatenate(
            (self.cartesian_coord(), self.cartesian_momentum()), axis=0)
